src/LibChatMessage.lua

```diff
@@ -1,549 +1,622 @@
+
+local lib = {}
+_G[LIB_IDENTIFIER] = lib
+
+local TAG_FORMAT = "[%s]"
+local COLOR_FORMAT = "|c%s%s|r"
+local MESSAGE_TEMPLATE = "%s %s"
+local SYSTEM_TAG = TAG_FORMAT:format(GetString("SI_CHATCHANNELCATEGORIES", CHAT_CATEGORY_SYSTEM))
+
+local TIME_FORMAT_AUTO = "[%X]"
+local TIME_FORMAT_12 = "[%I:%M:%S %p]"
+local TIME_FORMAT_24 = "[%T]"
+local TIME_FORMATS = { TIME_FORMAT_AUTO, TIME_FORMAT_12, TIME_FORMAT_24 }
+local TIME_FORMAT_MAPPING = {
+    ["auto"] = TIME_FORMAT_AUTO,
+    ["12h"] = TIME_FORMAT_12,
+    ["24h"] = TIME_FORMAT_24,
+}
+local REVERSE_TIME_FORMAT_MAPPING = {}
+for label, format in pairs(TIME_FORMAT_MAPPING) do
+    REVERSE_TIME_FORMAT_MAPPING[format] = label
+end
+
+local TAG_PREFIX_OFF = 1
+local TAG_PREFIX_LONG = 2
+local TAG_PREFIX_SHORT = 3
+
+local TIMESTAMP_INDEX = 1
+local MAX_HISTORY_LENGTH = 10000
+local TRIMMED_HISTORY_LENGTH = 9000
+
+local strlower = string.lower
+local tconcat = table.concat
+local osdate = os.date
+local GetTimeStamp = GetTimeStamp
+local CHAT_ROUTER = CHAT_ROUTER
+
+lib.defaultSettings = {
+    version = 1,
+    timePrefixEnabled = false,
+    timePrefixOnRegularChat = true,
+    timePrefixFormat = TIME_FORMAT_AUTO,
+    tagPrefixMode = TAG_PREFIX_LONG,
+    historyEnabled = false,
+    historyMaxAge = 3600,
+}
+lib.chatHistory = {}
+lib.chatHistoryActive = true
+
+lib.registeredChatLinks = {}
+
+-- internal functions
+
+local function GetFormattedTime(timeStamp)
+    return osdate(lib.settings.timePrefixFormat, timeStamp)
+end
+
+local function GetTimeStampForEvent()
+    if(lib.nextEventTimeStamp) then
+        return lib.nextEventTimeStamp, true
+    end
+    return GetTimeStamp(), false
+end
+
+local function StoreChatEvent(timeStamp, type, ...)
+    if(not lib.chatHistoryActive) then return end
+    local chatHistory = lib.chatHistory
+    chatHistory[#chatHistory + 1] = {timeStamp, type, ...}
+    if(#chatHistory > MAX_HISTORY_LENGTH) then
+        local newHistory = {}
+        for i = #chatHistory - TRIMMED_HISTORY_LENGTH, #chatHistory do
+            newHistory[#newHistory + 1] = chatHistory[i]
+        end
+        chatHistory = newHistory
+    end
+end
+
+local function ApplyTimeAndTagPrefix(formattedEventText, targetChannel, fromDisplayName, rawMessageText, timeStamp)
+    if(formattedEventText) then
+        if(lib.settings and lib.settings.tagPrefixMode ~= TAG_PREFIX_OFF) then
+            formattedEventText = MESSAGE_TEMPLATE:format(SYSTEM_TAG, formattedEventText)
+        end
+        if(lib.settings and lib.settings.timePrefixEnabled) then
+            formattedEventText = MESSAGE_TEMPLATE:format(GetFormattedTime(timeStamp), formattedEventText)
+        end
+    end
+    return formattedEventText, targetChannel, fromDisplayName, rawMessageText
+end
+
+-- chat system hooks
+local messageFormatters = CHAT_ROUTER:GetRegisteredMessageFormatters()
+local newFormatter = {}
+do
+    local redirect = {}
+    setmetatable(redirect, newFormatter)

```

Comment:
I think you mixed up the order of parameters? the real table is first and metatable should go second, otherwise this code doesn't make any sense as redirect isn't accessed anywhere